# Load Data

Pada tahap pertama, kita akan memuat data yang disimpah di google drive.

In [5]:
#Memuat file dari google drive
from google.colab import drive
drive.mount('/content/drive')
base_dir='/content/drive/MyDrive/Colab_Notebooks/data-train/'

Mounted at /content/drive


# Data Understanding

Pada tahap ini, kita akan melakukan pemahaman tentang data yang akan kita pakai sebelum membuat sistem komendasi untuk film berdasarkan sinopsis.

In [6]:
#Membaca file csv dengan library Pandas
import pandas as pd
df = pd.read_csv(base_dir+'/movies_metadata.csv')
df

/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py:3326: DtypeWarning: Columns (10) have mixed types.Specify dtype option on import or set low_memory=False.
  exec(code_obj, self.user_global_ns, self.user_ns)


adult                              belongs_to_collection    budget  \
0      False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1      False                                                NaN  65000000   
2      False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   
3      False                                                NaN  16000000   
4      False  {'id': 96871, 'name': 'Father of the Bride Col...         0   
...      ...                                                ...       ...   
45461  False                                                NaN         0   
45462  False                                                NaN         0   
45463  False                                                NaN         0   
45464  False                                                NaN         0   
45465  False                                                NaN         0   

                                                  genres  \
0      [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
1      [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   
2      [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   
3      [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...   
4                         [{'id': 35, 'name': 'Comedy'}]   
...                                                  ...   
45461  [{'id': 18, 'name': 'Drama'}, {'id': 10751, 'n...   
45462                      [{'id': 18, 'name': 'Drama'}]   
45463  [{'id': 28, 'name': 'Action'}, {'id': 18, 'nam...   
45464                                                 []   
45465                                                 []   

                                   homepage      id    imdb_id  \
0      http://toystory.disney.com/toy-story     862  tt0114709   
1                                       NaN    8844  tt0113497   
2                                       NaN   15602  tt0113228   
3                                       NaN   31357  tt0114885   
4                                       NaN   11862  tt0113041   
...                                     ...     ...        ...   
45461  http://www.imdb.com/title/tt6209470/  439050  tt6209470   
45462                                   NaN  111109  tt2028550   
45463                                   NaN   67758  tt0303758   
45464                                   NaN  227506  tt0008536   
45465                                   NaN  461257  tt6980792   

      original_language               original_title  \
0                    en                    Toy Story   
1                    en                      Jumanji   
2                    en             Grumpier Old Men   
3                    en            Waiting to Exhale   
4                    en  Father of the Bride Part II   
...                 ...                          ...   
45461                fa                      رگ خواب   
45462                tl          Siglo ng Pagluluwal   
45463                en                     Betrayal   
45464                en          Satana likuyushchiy   
45465                en                     Queerama   

                                                overview  ... release_date  \
0      Led by Woody, Andy's toys live happily in his ...  ...   1995-10-30   
1      When siblings Judy and Peter discover an encha...  ...   1995-12-15   
2      A family wedding reignites the ancient feud be...  ...   1995-12-22   
3      Cheated on, mistreated and stepped on, the wom...  ...   1995-12-22   
4      Just when George Banks has recovered from his ...  ...   1995-02-10   
...                                                  ...  ...          ...   
45461        Rising and falling between a man and woman.  ...          NaN   
45462  An artist struggles to finish his work while a...  ...   2011-11-17   
45463  When one of her hits goes wrong, a professiona...  ...   2003-08-01   
45464  In a small town live two brothers, one a minis...  ...   1917-10-21   
45465  50 years after decriminalisati

Kita memiliki 24 attribute data dimana masing-masing memiliki 45466 sample. Sekarang kita akan melihat tipe data tiap attribute.

In [7]:
#Fungsi info() untuk melihat tipe data
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45466 entries, 0 to 45465
Data columns (total 24 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   adult                  45466 non-null  object 
 1   belongs_to_collection  4494 non-null   object 
 2   budget                 45466 non-null  object 
 3   genres                 45466 non-null  object 
 4   homepage               7782 non-null   object 
 5   id                     45466 non-null  object 
 6   imdb_id                45449 non-null  object 
 7   original_language      45455 non-null  object 
 8   original_title         45466 non-null  object 
 9   overview               44512 non-null  object 
 10  popularity             45461 non-null  object 
 11  poster_path            45080 non-null  object 
 12  production_companies   45463 non-null  object 
 13  production_countries   45463 non-null  object 
 14  release_date           45379 non-null  object 
 15  re

Perhatikan bahwa kita sebagian besar memiliki tipe data bertipe object, kecuali revenue, runtime, vote_average, dan vote_count. Sekarang kita akan melihat seberapa banyak data yang unik untuk memeriksa apakah data memiliki duplikat atau missing value.

In [8]:
#Memeriksa data untuk attribute tittle
print('Banyak data: ', len(df.title.unique()))
print('Judul Film ', df.title.unique())

Banyak data:  42278
Judul Film  ['Toy Story' 'Jumanji' 'Grumpier Old Men' ... 'Century of Birthing'
 'Satan Triumphant' 'Queerama']


In [9]:
#Memeriksa data untuk attribute overview
print('Banyak data: ', len(df.overview.unique()))
print('Judul Film ', df.overview.unique())

Banyak data:  44308
Judul Film  ["Led by Woody, Andy's toys live happily in his room until Andy's birthday brings Buzz Lightyear onto the scene. Afraid of losing his place in Andy's heart, Woody plots against Buzz. But when circumstances separate Buzz and Woody from their owner, the duo eventually learns to put aside their differences."
 "When siblings Judy and Peter discover an enchanted board game that opens the door to a magical world, they unwittingly invite Alan -- an adult who's been trapped inside the game for 26 years -- into their living room. Alan's only hope for freedom is to finish the game, which proves risky as all three find themselves running from giant rhinoceroses, evil monkeys and other terrifying creatures."
 "A family wedding reignites the ancient feud between next-door neighbors and fishing buddies John and Max. Meanwhile, a sultry Italian divorcée opens a restaurant at the local bait shop, alarming the locals who worry she'll scare the fish away. But she's less i

Perhatikan bahwa kita memiliki jumlah data yang berbeda tiap attribute, yang artinya data kita memiliki missing value yang harus kita olah terlebih dahulu.

# Data Preparation

Pada tahap ini, kita akan mengolah dan memilah data pada attribute yang akan dipakai. Karena kita akan membuat sistem rekomendasi berdasarkan sinopsis maka kita hanya perlu dua attribute yang akan dipakai, yaitu tittle dan overview. Teknik yang akan dipakai juga tidak banya karena attribute yang dipakai hanya dua dan memiliki tipe data object.

Pertama kita akan membuang attribute yang tidak dibutuhkan dalam pembuatan sistem rekomendasi ini.

In [10]:
#Mmebuang attribute yang tidak dipakai
df.drop(['adult', 'belongs_to_collection', 'budget', 'genres', 'homepage', 'id', 'imdb_id', 'original_language', 'original_title', 
                'release_date', 'revenue', 'runtime', 'spoken_languages', 'status', 'tagline', 'video', 'vote_average', 'vote_count', 
         'popularity', 'poster_path', 'production_companies', 'production_countries'], axis=1, inplace=True)

In [11]:
df.head()

overview  \
0  Led by Woody, Andy's toys live happily in his ...   
1  When siblings Judy and Peter discover an encha...   
2  A family wedding reignites the ancient feud be...   
3  Cheated on, mistreated and stepped on, the wom...   
4  Just when George Banks has recovered from his ...   

                         title  
0                    Toy Story  
1                      Jumanji  
2             Grumpier Old Men  
3            Waiting to Exhale  
4  Father of the Bride Part II

Kemudian, kita akan memeriksa jumlah missing value pada masing-masing attribute.

In [12]:
#Banyak missing value tiap attribute
df.isnull().sum()

overview    954
title         6
dtype: int64

Perhatikan bahwa terdapat missing value sebanyak 954 untuk overview dan tittle sebanyak 6. Kita bisa melakukan dropping mengingat jumlah data yang kita pakai adalah puluhan ribu dan missing value yang sedikit. Alasan lain untuk melakukan dropping adalah sample tiap attribute adalah unik sehingga kita tidak bisa mengolahnya dengan pendekatan statistik seperti modus.

Karena ditemukan missing value yang sedikit maka kita bisa lakukan pembuangan sampel yang mengandung missing value.

In [13]:
#Membuang sampel yang memuat missing value
df1 = df.dropna()

In [14]:
#Memeriksa apakah masih ada missing value
df1.isnull().sum()

overview    0
title       0
dtype: int64

Perhatikan bahwa missing value pada masing-masing attribute sudah tidak ada. Oleh karena itu, kita bisa melakukan tahapan selanjutnya, yaitu Modelling & Result

# Modelling & Result

Pada tahap ini kita akan membuat sistem rekomendasi berdasarkan sinopsis dimana kita akan membuat kelas untuk sistem tersebut yang berisikan tiga fungsi, yaitu __init__() , fit(), dan rekomendasi().

__init__() : fungsi inisialisasi untuk memasukkan nilai awal, seperti data dan konten kolom yang akan digunakan 

fit() : fungsi untuk encoding kolom yang memuliki sampel bertipe object.

rekomendasi() : fungsi untuk melakukan ranking dengan menggunakan metriks cosine disantce.

In [15]:
from sklearn.feature_extraction.text import CountVectorizer
from nltk.tokenize import word_tokenize
from sklearn.metrics.pairwise import cosine_distances

#Pembuatan sistem rekomendasi berdasarkan sinopsis
#pembuatan class yang berisikan 3 fungsi
class SistemRekomendasi : 
  #fungsi inisialisasi
  def __init__(self, data, konten_kol):
    self.df = data
    self.konten_kol = konten_kol
    self.encoder = None
    self.bank = None
  
  #fungsi untuk encoding sample pada kolom overview
  def fit(self):
    self.encoder = CountVectorizer(stop_words = 'english', tokenizer = word_tokenize)
    self.bank = self.encoder.fit_transform(self.df[self.konten_kol])

  #Fungsi untuk melakukan rekomendasi berdasarkan metriks cosine distance
  def rekomendasi(self, idx, top_n=10):
    konten = df.loc[idx, self.konten_kol]
    kode = self.encoder.transform([konten])
    jarak = cosine_distances(kode, self.bank)
    rek_idx = jarak.argsort()[0, 1:(top_n + 1)]
    return self.df.loc[rek_idx]

In [16]:
import nltk
nltk.download('punkt')

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.


True

In [17]:
#Mencoba sistem rekomendasi
#untuk merekomendasikan film
#yang sam dengan jumanji
rekom = SistemRekomendasi(df1, konten_kol = 'overview')
rekom.fit()
rekom.rekomendasi(1) #index 1 : Jumanji

overview  \
28733  A young girl is brutally murdered somewhere in...   
43666  A hunting vacation goes horribly awry for 4 fr...   
40777  Novák cannot sleep. Not to disturb his girlfri...   
19597  The son of Torren learns of his heritage, goes...   
43463  The nano-plague that poisoned Earth’s water su...   
41058  The story of a day in the life of a lonely, se...   
39204  Elena (Kasia Smutniak) and Antonio (Francesco ...   
15445  Charles is a young provincial coming up to Par...   
38085  After learning the identity of the man who kil...   
13585  In this government-suggested sequel, Sugata ag...   

                          title  
28733          Who Saw Her Die?  
43666                 Buck Wild  
40777        My Night, Your Day  
19597  Ator, the Fighting Eagle  
43463        Shockwave Darkside  
41058        The Telephone Book  
39204     Fasten Your Seatbelts  
15445               Les Cousins  
38085                 Black Sun  
13585  Sanshiro Sugata Part Two

Hasil diatas merupakan hasil dari top-10 rekomendasi film berdasarkan sinopsis dari film Jumanji. Selanjutnya kita akan melakukan evaluasi apakah dengan rekomendasi berdasarkan sinopsis sistem dikatakan baik atau tidak.

#Evaluaiton

Pada tahap ini kita melakukan evaluasi terhadap sistem rekomendasi sudah dipakai. Tahapannya sama dengan kita membuat sistem sebelumnya. Pertama, kita akan melakukan transformasi terhadap attribute overview dengan fungsi CountVectorizer

In [18]:
#Transformasi data dengan CountVectorizer untuk seluruh sample pada kolom overview
bow = CountVectorizer(stop_words = "english", tokenizer = word_tokenize)
bank = bow.fit_transform(df1.overview)

In [19]:
#Memeriksa dan membuat variabel content untuk dilakukan evaluasi
idx = 2 # Film Grumpier Old Man
cont = df.loc[idx, "overview"]
cont

"A family wedding reignites the ancient feud between next-door neighbors and fishing buddies John and Max. Meanwhile, a sultry Italian divorcée opens a restaurant at the local bait shop, alarming the locals who worry she'll scare the fish away. But she's less interested in seafood than she is in cooking up a hot time with Max."

In [20]:
#Tranformasi data untuk sinopsis pada film Grumpier Old Man
code = bow.transform([cont])
code

<1x89429 sparse matrix of type '<class 'numpy.int64'>'
	with 34 stored elements in Compressed Sparse Row format>

Hasil encoding sebelumnya akan dijadikan evaluasi dengan menggunakan cosine distance untuk melihat jarak antar vektor tiap film berdasarkan sudut yang dibentuk. Disini kita bisa melihat sudut yang dibentuk antar vektor film Grumpier Old Man dengan film yang lainnya.

In [21]:
#Penghitungan jarak berdasarkan sudut yang dibentuk antar film
jarak = cosine_distances(code, bank)
jarak

array([[0.70198022, 0.73145692, 0.        , ..., 0.92284833, 0.7225617 ,
        0.76668587]])

In [22]:
#Pengurutan berdasrkan nilai cosine distance
rekom = jarak.argsort()[0, 1:11]
rekom

array([13277, 36267,  8004, 24983, 30928,  9090,  4526, 20813, 10769,
       12245])

In [23]:
#Melihat hasil dalam bentuk Dataframe
df.loc[rekom]

overview  \
13277  A woman is taken along with her mother in 1910...   
36267  ZERO is an 0 sucking in a young deserter, a ca...   
8004   Fictional account of French artist Henri de To...   
24983  Sam, a soldier who had served in Afghanistan a...   
30928  Based on the hit song that became a worldwide ...   
9090   Angel, an exterminator recently released from ...   
4526   Romero is a compelling and deeply moving look ...   
20813  College students exploring the Ozark Mountains...   
10769  In this reworking of "No, No, Nanette," wealth...   
12245  A documentary that examines the psychopatholog...   

                                    title  
13277                   The House of Sand  
36267               Zero. Lilac Lithuania  
8004                         Moulin Rouge  
24983                         Amira & Sam  
30928  Grandma Got Run Over by a Reindeer  
9090                                Earth  
4526                               Romero  
20813                         Albino Farm  
10769                         Tea for Two  
12245                     Suicide Killers

Selanjutnya, kita akan mengevaluasi sistem rekomendasi yang sudah kita buat dengan menggunakan nDCG score. nDCG adalah rasio skor DCG peserta atas skor DCG peringkat ideal. Score yang diperoleh dari nDCG berada diantara 0 dan 1. Semakin besar nilai socrenya maka sistem kita berhasil melakukan ranking terhadap suatu item.

Pada tahap ini kita akan mambuat array yang berisi top-10 rekomendasi dari film yang sudah kita rujuk sebagai nilai prediksi dan array yang berisi top 10 rekomendasi berdasarkan penalaran kita atau ranking dari google sebagai nilai aktual.

In [24]:
import numpy as np
#Prediksi 
pred = [[cosine_distances(code, bank[13277])[0][0],
         cosine_distances(code, bank[36267])[0][0],
         cosine_distances(code, bank[8004])[0][0],
         cosine_distances(code, bank[24983])[0][0],
         cosine_distances(code, bank[30928])[0][0],
         cosine_distances(code, bank[9090])[0][0],
         cosine_distances(code, bank[4526])[0][0],
         cosine_distances(code, bank[20813])[0][0],
         cosine_distances(code, bank[10769])[0][0],
         cosine_distances(code, bank[12245])[0][0],
         ]]

np.asarray(pred)

array([[0.48968964, 0.50184163, 0.50626302, 0.51289198, 0.51950617,
        0.52233485, 0.52254045, 0.52565835, 0.52735733, 0.52835002]])

In [25]:
#Aktual
aktual = [[cosine_distances(code, bank[334])[0][0], #While You Were Sleeping
          cosine_distances(code, bank[5483])[0][0], #Sweet Home Alabama
          cosine_distances(code, bank[907])[0][0], #Father of the Bride
          cosine_distances(code, bank[2579])[0][0], #Big Daddy
          cosine_distances(code, bank[4492])[0][0], #Look Who's Talking
          cosine_distances(code, bank[248])[0][0], # I.Q.
          cosine_distances(code, bank[328])[0][0], #Tommy Boy
          cosine_distances(code, bank[677])[0][0], #Mrs. Winterbourne
          cosine_distances(code, bank[2609])[0][0], #Runaway Bride
          cosine_distances(code, bank[6028])[0][0], # Dennis the Menace
          ]]

np.asarray(aktual)

array([[0.68722838, 0.69166508, 0.7763932 , 0.72916667, 0.76585355,
        0.72304415, 0.58549759, 0.84613564, 0.64761926, 0.76851852]])

Kita akan menghitubg score nDCG dengan menggunakan library sklearn.metrics dengan fungsi ndcg_score.

In [26]:
#Melihat seberapa baik sistem kira untuk ranking dengan metriks nDCG
from sklearn.metrics import ndcg_score
ndcg_score(np.asarray(aktual), np.asarray(pred))

0.9701417955219164

Perhatikan bahwa nilai yang dihasilkan mendekati satu, artinya sistem rekomendasi yang kita buat melakukan ranking dengan baik berdasarkan nilai yang dikembalikan oleh fungsi similarity yang kita deklarasikan. Namun, Meskipun hasil yang diberikan baik berdasarkan score yang didapat, hasil realita tidak sesuai yang diinginkan. Misal, While You Were Sleeping adalah film romantis yang mana sama dengan film yang kita rujul tetapi sistem tidak merekomendasikan hal tersebut karena sistem merekomendasikan berdasarkan sinopsis saja.